akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardedDaemonProcessImpl.scala

```diff
@@ -0,0 +1,167 @@
+
+  def idFromEntityId(entityId: EntityId, name: String): Int =
+    stringIdFromEntityId(entityId, name).toInt
+
+  def stringIdFromEntityId(entityId: EntityId, name: String): String =
+    entityId.drop(name.length + 1)
+
+  final class MessageExtractor[T](name: String) extends ShardingMessageExtractor[ShardingEnvelope[T], T] {
+    def entityId(message: ShardingEnvelope[T]): String = message match {
+      case ShardingEnvelope(id, _) => id
+    }
+
+    def shardId(entityId: String): String = stringIdFromEntityId(entityId, name)
+
+    def unwrapMessage(message: ShardingEnvelope[T]): T = message.message
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] final class ShardedDaemonProcessImpl(system: ActorSystem[_])
+    extends javadsl.ShardedDaemonProcess
+    with scaladsl.ShardedDaemonProcess {
+
+  import ShardedDaemonProcessImpl._
+
+  def init[T](name: String, numberOfInstances: Int, behaviorFactory: Int => Behavior[T])(
+      implicit classTag: ClassTag[T]): Unit = {
+    init(name, numberOfInstances, behaviorFactory, ShardedDaemonProcessSettings(system), None)(classTag)
+  }
+
+  def init[T](
+      name: String,
+      numberOfInstances: Int,
+      behaviorFactory: Int => Behavior[T],
+      settings: ShardedDaemonProcessSettings,
+      stopMessage: Option[T])(implicit classTag: ClassTag[T]): Unit = {
+
+    val entityTypeKey = EntityTypeKey[T](s"sharded-daemon-process-$name")
+
+    // One shard per actor identified by the numeric id encoded in the entity id
+    val numberOfShards = numberOfInstances
+    val entityIds = (0 to numberOfInstances).map(idAndNameToEntityId(_, name))
+
+    val shardingSettings = {
+      val settingsFromConfig =
+        ClusterShardingSettings.fromConfig(
+          // defaults in akka.cluster.sharding but allow overrides specifically for actor-set
+          system.settings.config.getConfig("akka.cluster.sharded-daemon-process.sharding"))
+
+      new ClusterShardingSettings(
+        numberOfShards,
+        settingsFromConfig.role,
+        settingsFromConfig.dataCenter,
+        false, // remember entities disabled
+        "",
+        "",
+        Duration.Zero, // passivation disabled
+        settingsFromConfig.shardRegionQueryTimeout,
+        StateStoreModeDData,
+        settingsFromConfig.tuningParameters,
+        settingsFromConfig.coordinatorSingletonSettings)
+    }
+
+    val entity = Entity(entityTypeKey)(ctx => behaviorFactory(idFromEntityId(ctx.entityId, name)))
+      .withSettings(shardingSettings)
+      .withMessageExtractor(new MessageExtractor(name))
+
+    val entityWithStop = stopMessage match {
+      case Some(stop) => entity.withStopMessage(stop)
+      case None       => entity
+    }
+
+    val shardingRef = ClusterSharding(system).init(entityWithStop)
+
+    system.systemActorOf(
+      KeepAlivePinger(settings, entityIds.toSet, shardingRef),
+      s"shardedDaemonProcessKeepAlive-$name")
```

Comment:
in ShardingSettings there is a `role`, which we can look at there to only start the pinger on nodes with the given role. The whole init can be ignored if not correct role.